### PR TITLE
Merge Wishbone branch onto master

### PR DESCRIPTION
**Pull request by chiggs**
Currently the work by @mkreider is on the `wishbone` branch but should be merged onto `master`.

Use this PR to discuss any outstanding work and decide when to merge.

cc @jeremyherbert @Martoni

### CODE REVIEW

**Comment by chiggs** [comment]
Note this is a continuation of pull request #335.

**Comment by Martoni** [comment]
I don't understand the exact problem that block this merge. Does somebody has reproduced the problem on his computer ?

**Comment by imphil** [comment]
Closing now in favor of  #862, please follow up there.

File: cocotb/drivers/wishbone.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+                for res, aux in zip(self._res_buf, self._aux_buf):
+                    res.datwr       = aux.datwr
+                    res.sel         = aux.sel
+                    res.adr         = aux.adr
+                    res.waitIdle    = aux.waitIdle
+                    res.waitStall   = aux.waitStall
+                    res.waitAck    -= aux.ts
+                    result.append(res)
+
+            raise ReturnValue(result)
```
Comment (eric-wieser) [inline cocotb/drivers/wishbone.py]: This crashes with `UnboundLocalError` if `len(arg) < 1`

File: cocotb/drivers/wishbone.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        else:
+            raise TestFailure("Sorry, argument must be a list of WBOp (Wishbone Operation) objects!")
+            raise ReturnValue(None)
```
Comment (eric-wieser) [inline cocotb/drivers/wishbone.py]: Unreachable

File: cocotb/drivers/wishbone.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        yield clkedge
+        if is_sequence(arg):
+            if len(arg) < 1:
+                self.log.error("List contains no operations to carry out")
+            else:
+                self._op_cnt = len(arg)
+                firstword = True
+                for op in arg:
+                    if not isinstance(op, WBOp):
+                        raise TestFailure("Sorry, argument must be a list of WBOp (Wishbone Operation) objects!")
```
Comment (eric-wieser) [inline cocotb/drivers/wishbone.py]: Should be `TypeError`

File: cocotb/monitors/wishbone.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+            
+
+class WishboneSlave(Wishbone):
+    """Wishbone slave
+    """
+    
+    def bitSeqGen(self, tupleGen):
+        while True: 
+            [highCnt, lowCnt] = tupleGen.next()
```
Comment (eric-wieser) [inline cocotb/monitors/wishbone.py]: Doesn't work in python 3

File: cocotb/monitors/wishbone.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+               bits.append(1)          
+            for i in range(0, lowCnt):
+               bits.append(0)
+            for bit in bits:
+                yield bit
+    
+    
+    def defaultTupleGen(self):
+        while True:        
+            yield int(0), int(1)
```
Comment (eric-wieser) [inline cocotb/monitors/wishbone.py]: Better spelt `itertools.repeat((0, 1))`

File: cocotb/monitors/wishbone.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        while True:        
+            yield int(0), int(1)      
+    
+    def defaultGen0(self):
+        while True:        
+            yield int(0)
+            
+    def defaultGen1(self):
+        while True:        
+            yield int(1)
```
Comment (eric-wieser) [inline cocotb/monitors/wishbone.py]: Better spelt `itertools.repeat(1)`

File: cocotb/monitors/wishbone.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+                yield bit
+    
+    
+    def defaultTupleGen(self):
+        while True:        
+            yield int(0), int(1)      
+    
+    def defaultGen0(self):
+        while True:        
+            yield int(0)
```
Comment (eric-wieser) [inline cocotb/monitors/wishbone.py]: Better spelt `itertools.repeat(0)`

File: tests/test_cases/test_wishbone/test_wishbone.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+from cocotb.drivers.wishbone import WishboneMaster
+
+class GPIOTB(object):
+
+    def __init__(self, dut, debug=False):
+        self.dut = dut
+        self.log = logging.getLogger("cocotb.gpiotb")
+
+        # 
+_signals = ["cyc", "stb", "we", "sel", "adr", "datwr", "datrd", "ack"]
```
Comment (eric-wieser) [inline tests/test_cases/test_wishbone/test_wishbone.py]: Dead code

**Comment by Fatsie** [comment]
I am working on it, mainly also implementing pipelined interface.
See https://gitlab.com/Chips4Makers/cocotb/tree/dev_wishbone.

**Comment by mkreider** [comment]
@eric-wieser Thanks for the reply. I was disappointed by the standstill in combination with the lack of feedback. You are the first person who pointed what exactly needs to be improved and gave tips how, the way I expect good open source collaborations to work. If I find the time I'll do necessary fixup or support @Fatsie with it if he likes the help.
However, I'd like some clarification on 
> even if we did want it  

first - this says you are not interested in using this code? If so, I can save myself some work, so can @Fatsie

**Comment by mkreider** [comment]
@Fatsie Cool! Glad someone found it that useful. Just one thing I don't understand - the old version already supported pipelined mode, what do you mean by 

> implementing pipelined interface

**Comment by eric-wieser** [comment]
> first - this says you are not interested in using this code?

There is a leaning to this sort of code living outside the cocotb main repository, which is being tracked in #1141 - the rationale being that bus protocols are rather specialist, and the people suited to maintaining C++ simulator interfaces and asynchronous schedulers are unlikely to be able to make judgement calls on protocols they've never used before.

**Comment by Fatsie** [comment]
@mkreider 
OK, the patch is under development and not clean yet.
I needed to fix the code to work with my pipelined Wishbone code but it can mean different things:
* My RTL code is wrong.
* I had to fix a problem on something I messed up myself in the cocotb code.
* I stumbled on a genuine bug in the code.

**Comment by OesterKweker** [comment]
I'm interested in wishbone driver and monitors. Difficult to track what the status and the plans are.

**Comment by Martoni** [comment]
@OesterKweker you can take a look at cocotbext-wishbone module [here](https://github.com/wallento/cocotbext-wishbone).

**Comment by OesterKweker** [comment]
@martoni great, thanks a lot!

**Pull request closed without merging**